pennylane/optimize/qnspsa.py

```diff
@@ -0,0 +1,382 @@
+            (-self.stepsize * grad_vec + np.matmul(self.metric_tensor, params_vec)),
+        )
+
+        params_split_indices = []
+        tmp = 0
+        for param in params:
+            tmp += param.size
+            params_split_indices.append(tmp)
+        new_params = np.split(new_params_vec, params_split_indices)
+        new_params_reshaped = [new_params[i].reshape(params[i].shape) for i in range(len(params))]
+        return new_params_reshaped
+
+    def _get_spsa_grad_tapes(self, cost, args, kwargs):
+        dirs = []
+        args_plus = list(args)
+        args_minus = list(args)
+        for index, arg in enumerate(args):
+            if not getattr(arg, "requires_grad", False):
+                continue
+            direction = self.rng.choice([-1, 1], size=arg.shape)
+
+            dirs.append(direction)
+            args_plus[index] = arg + self.finite_diff_step * direction
+            args_minus[index] = arg - self.finite_diff_step * direction
+
+        cost.construct(args_plus, kwargs)
+        tape_plus = cost.tape.copy(copy_operations=True)
+        cost.construct(args_minus, kwargs)
+        tape_minus = cost.tape.copy(copy_operations=True)
+        return [tape_plus, tape_minus], dirs
+
+    def _update_tensor(self, tensor_raw):
+        tensor_avg = self._get_tensor_moving_avg(tensor_raw)
+        tensor_regularized = self._regularize_tensor(tensor_avg)
+        self.metric_tensor = tensor_regularized
+        self.k += 1
+
+    def _get_tensor_tapes(self, cost, args, kwargs):
+        dir1_list = []
+        dir2_list = []
+        args_list = [list(args) for _ in range(4)]
+
+        for index, arg in enumerate(args):
+            if not getattr(arg, "requires_grad", False):
+                continue
+            dir1 = self.rng.choice([-1, 1], size=arg.shape)
+            dir2 = self.rng.choice([-1, 1], size=arg.shape)
+
+            dir1_list.append(dir1.reshape(-1))
+            dir2_list.append(dir2.reshape(-1))
+
+            args_list[0][index] = arg + self.finite_diff_step * (dir1 + dir2)
+            args_list[1][index] = arg + self.finite_diff_step * dir1
+            args_list[2][index] = arg + self.finite_diff_step * (-dir1 + dir2)
+            args_list[3][index] = arg - self.finite_diff_step * dir1
+        dir_vecs = (np.concatenate(dir1_list), np.concatenate(dir2_list))
+        tapes = []
+        for args_finite_diff in args_list:
+            tapes.append(self._get_overlap_tape(cost, args, args_finite_diff, kwargs))
+        return tapes, dir_vecs
+
+    def _get_overlap_tape(self, cost, args1, args2, kwargs):
+        op_forward = self._get_operations(cost, args1, kwargs)
+        op_inv = self._get_operations(cost, args2, kwargs)
+
+        with qml.tape.QuantumTape() as tape:
+            for op in op_forward:
+                qml.apply(op)
+            for op in reversed(op_inv):
+                op.adjoint()
+            qml.probs(wires=cost.tape.wires.labels)
```

Comment:
Good call! This approach will not work for ``qml.QubitStateVector``, as its adjoint is not defined yet. It is a bit unclear if I can make some adjustment to make the code compatible with ``qml.QubitStateVector``. The naive thought is to extract the operation decomposition from it and then apply adjoint. But I kind of think this is out of the scope of this optimizer. Instead, to me, changes should happen in the adjoint method. Let me know what is your preference.

Reply:
It might be better to calculate the overlap another way. The most intuitive is to create two tapes that produce the final states, then return the square of absolute value of the inner product between them. (There's a similar formula for density matrices too if necessary.)

Reply:
I am a bit confused: how would I compute this inner product of two quantum states in a physical way? Are you thinking of using ``2 * n_qubit`` registers to do so?

Reply:
Yes, I meant using `2 * n_qubit` registers to prepare the two states (that is, return `qml.state()` from the tapes). The inner product computation can then be done classically. 

On further investigation, it's actually a bug that `qml.QubitStateVector` doesn't support an adjoint: https://github.com/PennyLaneAI/pennylane/issues/2666. I'm not sure when this is going to be fixed, but I'd say it's probably fine to leave this code as-is right now. So never mind 😃 .

Reply:
So generally there are two approaches to do this state overlap:

- doubling the qubit register number
- doubling the circuit depth

The 2nd one is more preferable to me to avoid memory overhead for simulators. I am rather sure there could be circuit cutting or other techniques to decouple the two group of registers, but on top of my head I don't have a clue on how to do it compatibly for both simulators and quantum devices. For now will keep the solution, but will read into this sometime.